hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java

```diff
@@ -390,22 +391,55 @@ void preInsertInMemoryValueGeneration(Object[] fields, Object object,
 		} );
 
 		SessionFactoryImplementor factory = session.getFactory();
-		Dialect dialect = factory.getJdbcServices().getDialect();
 		ReactiveConnection connection = getReactiveConnection( session );
 		if ( factory.getSessionFactoryOptions().isGetGeneratedKeysEnabled() ) {
-			//TODO: wooooo this is awful ... I believe the problem is fixed in Hibernate 6
-			if ( dialect instanceof PostgreSQL81Dialect) {
-				sql = sql + " returning " + delegate().getIdentifierColumnNames()[0];
-			}
-			return connection.updateReturning( sql, params ).thenApply( id -> id );
+			return connection.updateReturning( checkSql( sql ), params )
+					.thenApply( this::castToIdentityType );
 		}
 		else {
 			//use an extra round trip to fetch the id
 			return connection.update( sql, params )
 					.thenCompose( v -> connection.selectLong( delegate().getIdentitySelectString(), EMPTY_OBJECT_ARRAY ) )
-					.thenApply( id -> id );
+					.thenApply( this::castToIdentityType );
 		}
+	}
 
+	/**
+	 * Queries used to insert a new element and retrieve the id in one go require
+	 * some changes
+	 */
+	default String checkSql(String sql) {
+		Dialect dialect = getFactory().getJdbcServices().getDialect();
+		//TODO: wooooo this is awful ... I believe the problem is fixed in Hibernate 6
+		if ( dialect instanceof PostgreSQL81Dialect ) {
+			return sql + " returning " + delegate().getIdentifierColumnNames()[0];
+		}
+		if ( dialect instanceof DB2Dialect ) {
+			return "select " + delegate().getIdentifierColumnNames()[0] + " from NEW TABLE (" + sql + ")";
+		}
+		return sql;
+	}
+
+	/**
+	 * A simplified version of
+	 * {@link org.hibernate.id.IdentifierGeneratorHelper#getGeneratedIdentity(ResultSet, String, Type, Dialect)}.
+	 */
+	default Serializable castToIdentityType(Long generatedId) {
+		log.debugf( "Natively generated identity: %s", generatedId );
+		if ( generatedId == null ) {
+			throw new HibernateException( "The database returned no natively generated identity value" );
+		}
+		Class clazz = delegate().getIdentifierType().getReturnedClass();
+		if ( clazz == Long.class ) {
+			return generatedId;
+		}
+		if ( clazz == Integer.class ) {
+			return generatedId.intValue();
+		}
+		if ( clazz == Short.class ) {
+			return generatedId.shortValue();
+		}
```

Comment:
@DavideD this sorta duplicates the logic in `IdentifierGeneration.assignIdIfNecessary()`. I think we should find a way to unify both into one method.

Reply:
At the _very least_ it should be moved to a `static` method of `IdentifierGeneration`.

Reply:
Makes sense.

I actually noticed that, but I when I looked at `IdentifierGeneration` I didn't feel like this method would fit there.
I will give  a second look

Reply:
No, you are right. I will update the PR